lib/twitter_cldr/shared/locale.rb

```diff
@@ -0,0 +1,265 @@
+# encoding: UTF-8
+
+# Copyright 2012 Twitter, Inc
+# http://www.apache.org/licenses/LICENSE-2.0
+
+module TwitterCldr
+  module Shared
+    class Locale
+
+      class << self
+        # http://unicode.org/reports/tr35/tr35-9.html#Likely_Subtags
+        #
+        # 1. Make sure the input locale is in canonical form: uses the right
+        #    separator, and has the right casing.
+        #
+        # 2. Replace any deprecated subtags with their canonical values using
+        #    the <alias> data in supplemental metadata. Use the first value in
+        #    the replacement list, if it exists.
+        #
+        # 3. If the tag is grandfathered (see <variable id="$grandfathered"
+        #    type="choice"> in the supplemental data), then return it.
+        #
+        # 4. Remove the script code 'Zzzz' and the region code 'ZZ' if they
+        #    occur; change an empty language subtag to 'und'.
+        #
+        # 5. Get the components of the cleaned-up tag (language¹, script¹, and
+        #    region¹), plus any variants if they exist (including keywords).
+        def parse(locale_text)
+          locale_text.strip!
```

Comment:
Might be safer to do `locale_text = locale_text.strip` here in case someone keeps using that string outside of the method.